File: July/Day6/finding_pairs_with_a_certain_sum.cc
#include <vector>
#include <unordered_map>
using namespace std;

class FindSumPairs {
    vector<int> nums1;
    vector<int> nums2;
    unordered_map<int, int> freqMap;

public:
    FindSumPairs(vector<int>& nums1, vector<int>& nums2) {
        this->nums1 = nums1;
        this->nums2 = nums2;
        for (int num : nums2) {
            freqMap[num]++;
        }
    }

    void add(int index, int val) {
        int prevVal = nums2[index];
        nums2[index] += val;

        freqMap[prevVal]--;
        if (freqMap[prevVal] == 0) {
            freqMap.erase(prevVal);
        }
        freqMap[nums2[index]]++;
    }

    int count(int tot) {
        int cnt = 0;
        for (int num1 : nums1) {
            int num2 = tot - num1;
            if (freqMap.count(num2)) {
                cnt += freqMap[num2];
            }
        }
        return cnt;
    }
};

File: July/Day6/finding_pairs_with_a_certain_sum.java
import java.util.*;

class FindSumPairs {
    private int[] nums1;
    private int[] nums2;
    private Map<Integer, Integer> freqMap;

    public FindSumPairs(int[] nums1, int[] nums2) {
        this.nums1 = nums1.clone();
        this.nums2 = nums2.clone();
        freqMap = new HashMap<>();
        for (int num : nums2) {
            freqMap.put(num, freqMap.getOrDefault(num, 0) + 1);
        }
    }

    public void add(int index, int val) {
        int prevVal = nums2[index];
        nums2[index] += val;

        freqMap.put(prevVal, freqMap.get(prevVal) - 1);
        if (freqMap.get(prevVal) == 0) {
            freqMap.remove(prevVal);
        }

        freqMap.put(nums2[index], freqMap.getOrDefault(nums2[index], 0) + 1);
    }

    public int count(int tot) {
        int cnt = 0;
        for (int num1 : nums1) {
            int num2 = tot - num1;
            cnt += freqMap.getOrDefault(num2, 0);
        }
        return cnt;
    }
}
